# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

In [2]:
# URL for data
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [8]:
import pandas as pd

df = pd.read_csv(url)

# Dimension of data 
print(f"Shape of DF is: {df.shape}")


Shape of DF is: (4008, 11)


In [58]:
# Data types of columns
print("Data types of columns:\n") 
print(df.dtypes)

''' I think the data types for these columns are appropiate. I would suggest changing the data type of the Number of Open Complaints to a numeric type
and perhaps the Lifetime Value column'''

df['Customer Lifetime Value'] = df['Customer Lifetime Value'].str.replace('%', '').astype('float') / 100
df['Customer Lifetime Value'].describe()

Data types of columns:

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object


count     1068.000000
mean      7936.902593
std       6434.784103
min       2004.350700
25%       4034.079950
50%       5881.742350
75%       8962.871975
max      58166.553500
Name: Customer Lifetime Value, dtype: float64

In [59]:
# Number of unique values 
print("Unique values per column: \n")
print(df.nunique())

Unique values per column: 

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64


Given the count of unique values, I believe that the columns that hold categorical data are the following:
1. ST = the state of the Customer
2. Gender
3. Education
4. Vehicle Class
5. Policy Type



In [60]:
numerical_data = df.select_dtypes(include='number')
numerical_data.describe()

Customer Lifetime Value        Income  Monthly Premium Auto  \
count              1068.000000   1071.000000           1071.000000   
mean               7936.902593  39295.701214            193.234360   
std                6434.784103  30469.427060           1601.190369   
min                2004.350700      0.000000             61.000000   
25%                4034.079950  14072.000000             68.000000   
50%                5881.742350  36234.000000             83.000000   
75%                8962.871975  64631.000000            109.500000   
max               58166.553500  99960.000000          35354.000000   

       Total Claim Amount  
count         1071.000000  
mean           404.986909  
std            293.027260  
min              0.382107  
25%            202.157702  
50%            354.729129  
75%            532.800000  
max           2893.239678

Observations on Numerical Data
1. Income Column
    1. 50% of insurance holders claim to have annual incomes below the mean. The presence of higher incomes skews the value of the mean
    2. High standard deviation tells that there is a lot of variability with the datapoints. 
    3. There are insurance holders who claim to have no annual income

2. Monthly Premium Auto Column
    1. Lower standard deviation tells us that datapoints hover closer to the mean
    2. The lowest monthly auto payment is of $61.00
    3. 75% of customers pay a monthly payment for insurance below the mean. However, 
        looking at the maximum amount of monthly insurance payment, we can see that it 
        causes the mean to be skewed as well. 

3. Total Claim Amount Column
    1. Lower standard deviation tells us that datapoints hover closer to the mean
    2. 50% of insurance holders claim insurance amounts below the mean. 
    3. The least amount of claimed insurance by a customer is of $0.382107 


In [61]:
object_data = df.select_dtypes(include='object')
object_data.describe()

Customer      ST GENDER Education Number of Open Complaints  \
count      1071    1071    954      1071                      1071   
unique     1071       8      5         6                         6   
top     RB50392  Oregon      F  Bachelor                    1/0/00   
freq          1     320    457       324                       830   

          Policy Type  Vehicle Class  
count            1071           1071  
unique              3              6  
top     Personal Auto  Four-Door Car  
freq              780            576

Observations on Categorical Data

1. ST Column
    - Of a total of 8 states indicated by customers, the one where most of the customers live in is Oregon 
        with a frequency of **320** times

2. Gender Column
    - From 5 options to choose from, most customers from this dataset identify as Female with a frequency of **457** times. This indicates a 47% of the datapoints in this column.

3. Education Column
    - The most frequent education listed by our customers is a Bachelor's degree. Curiously this represents around 30% percent of the observations in this column. 

4. Policy Type
    - The most common Policy type is the Personal Auto policy with a frequency of **780** out of 1071 observations. 

5. Vehicle Class
    - The top value in the Vehicle Class column is the Four-Door-Car class with **576** out of 1071 observations.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [62]:
# Your code here
frequency_cities = df["ST"].value_counts()
frequency_cities = frequency_cities.sort_values(ascending=True)
print("Least common locations: ")
print(frequency_cities.head())

Least common locations: 
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: ST, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [63]:
# Your code here
policies = df["Policy Type"]

print("Number of policies sold by Type:")
print(policies.value_counts())

print("\nThe most sold policy type is:")
print(policies.value_counts()[:1])


Number of policies sold by Type:
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: Policy Type, dtype: int64

The most sold policy type is:
Personal Auto    780
Name: Policy Type, dtype: int64


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [64]:
# Your code here
personal_auto = df.loc[df["Policy Type"] == "Personal Auto"]
personal_auto

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
4     GA49547  Washington      M  High School or Below   
5     OC83172      Oregon      F              Bachelor   
...       ...         ...    ...                   ...   
1065  XD85577        Cali      M              Bachelor   
1066  TM65736      Oregon      M                Master   
1067  VJ51327        Cali      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1070  MY31220  California      F               College   

      Customer Lifetime Value   Income  Monthly Premium Auto  \
0                         NaN      0.0                1000.0   
1                   6979.5359      0.0                  94.0   
2                  12887.4317  48767.0                 108.0   
4                   5363.0765  36357.0                  68.0   
5                   8256.2978  62902.0                  69.0   
...                       ...      ...                   ...   
1065                2842.2669  69417.0                  73.0   
1066                3059.5503  38644.0                  78.0   
1067               20314.9976  63209.0                 102.0   
1068                3239.1247  16061.0                  88.0   
1070                8997.0402  54230.0                 112.0   

     Number of Open Complaints    Policy Type  Vehicle Class  \
0                       1/0/00  Personal Auto  Four-Door Car   
1                       1/0/00  Personal Auto  Four-Door Car   
2                       1/0/00  Personal Auto   Two-Door Car   
4                       1/0/00  Personal Auto  Four-Door Car   
5                       1/0/00  Personal Auto   Two-Door Car   
...                        ...            ...            ...   
1065                    1/1/00  Personal Auto  Four-Door Car   
1066                    1/1/00  Personal Auto  Four-Door Car   
1067                    1/2/00  Personal Auto            SUV   
1068                    1/0/00  Personal Auto  Four-Door Car   
1070                    1/0/00  Personal Auto   Two-Door Car   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
4              17.269323  
5             159.383042  
...                  ...  
1065           30.874869  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1070          537.600000  

[780 rows x 11 columns]

In [65]:
corporate_auto = df.loc[df["Policy Type"] == "Corporate Auto"]
corporate_auto

Customer          ST GENDER             Education  \
3     WW63253  California      M              Bachelor   
6     XZ87318      Oregon      F               College   
7     CF85061     Arizona      M                Master   
8     DY87989      Oregon      M              Bachelor   
12    DP39365  California    NaN                Master   
...       ...         ...    ...                   ...   
1045  WZ31900      Oregon      F              Bachelor   
1050  BJ53923     Arizona      M  High School or Below   
1052  FB23788      Oregon      M  High School or Below   
1056  PI47776      Oregon      F              Bachelor   
1059  YG44474      Oregon      M               College   

      Customer Lifetime Value   Income  Monthly Premium Auto  \
3                   7645.8618      0.0                 106.0   
6                   5380.8986  55350.0                  67.0   
7                   7216.1003      0.0                 101.0   
8                  24127.5040  14072.0                  71.0   
12                  8798.7970  77026.0                 110.0   
...                       ...      ...                   ...   
1045                8649.7006  94389.0                 107.0   
1050                2600.2721  51978.0                  66.0   
1052                8828.8350  86721.0                 111.0   
1056                2530.7051  89451.0                  63.0   
1059               14014.7213  54193.0                 117.0   

     Number of Open Complaints     Policy Type  Vehicle Class  \
3                       1/0/00  Corporate Auto            SUV   
6                       1/0/00  Corporate Auto  Four-Door Car   
7                       1/0/00  Corporate Auto  Four-Door Car   
8                       1/0/00  Corporate Auto  Four-Door Car   
12                      1/2/00  Corporate Auto  Four-Door Car   
...                        ...             ...            ...   
1045                    1/0/00  Corporate Auto            SUV   
1050                    1/0/00  Corporate Auto  Four-Door Car   
1052                    1/0/00  Corporate Auto            SUV   
1056                    1/0/00  Corporate Auto  Four-Door Car   
1059                    1/0/00  Corporate Auto            SUV   

      Total Claim Amount  
3             529.881344  
6             321.600000  
7             363.029680  
8             511.200000  
12            472.029737  
...                  ...  
1045           85.063708  
1050          144.782152  
1052          532.800000  
1056           61.769564  
1059          720.752945  

[234 rows x 11 columns]

In [66]:
print("The average income for customers under the Personal Auto policy type is:")
print(round(personal_auto["Income"].mean(), 2))

print("The average income for customers under the Corporate Auto policy type is:")
print(round(corporate_auto["Income"].mean(), 2))

The average income for customers under the Personal Auto policy type is:
38180.7
The average income for customers under the Corporate Auto policy type is:
41390.31


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [67]:
# Your code here
df["Total Claim Amount"].describe()

count    1071.000000
mean      404.986909
std       293.027260
min         0.382107
25%       202.157702
50%       354.729129
75%       532.800000
max      2893.239678
Name: Total Claim Amount, dtype: float64

In [68]:
top_25 = df["Total Claim Amount"].quantile(0.75)
high_policy_claims = df.loc[df["Total Claim Amount"] > top_25]

high_policy_claims.describe()

Customer Lifetime Value        Income  Monthly Premium Auto  \
count               264.000000    264.000000            264.000000   
mean               9086.868131  23677.344697            165.193182   
std                6387.677898  27013.483721            623.930992   
min                2287.596900      0.000000             63.000000   
25%                4723.957925      0.000000             99.000000   
50%                7781.768250  18807.000000            114.000000   
75%               10544.299275  42423.750000            133.250000   
max               38445.855900  99316.000000          10202.000000   

       Total Claim Amount  
count          264.000000  
mean           782.228263  
std            292.751640  
min            537.600000  
25%            606.521741  
50%            679.597985  
75%            851.400000  
max           2893.239678